pkg/util/podstore/podstore.go

```diff
@@ -0,0 +1,227 @@
+		})
+	return s
+}
+
+func NewPodStore(podInformer cache.SharedIndexInformer) *PodStore {
+	return NewPodStoreWithClock(podInformer, clock.RealClock{})
+}
+
+func (s *PodStore) onPodUpdate(oldObj interface{}, newObj interface{}) {
+	newPod, ok := newObj.(*corev1.Pod)
+	if !ok {
+		klog.ErrorS(nil, "Received unexpected object", "newObj", newObj)
+		return
+	}
+	s.pods.Update(newPod)
+	klog.V(4).InfoS("Process Pod Update Event", "Pod", klog.KObj(newPod))
+}
+
+func (s *PodStore) onPodCreate(obj interface{}) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	timeNow := s.clock.Now()
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		klog.ErrorS(nil, "Received unexpected object", "obj", obj)
+		return
+	}
+	err := s.pods.Add(pod)
+	if err != nil {
+		klog.ErrorS(err, "Error when adding pod to curPods")
+		return
+	}
+	switch pod.Status.Phase {
+	case corev1.PodPending:
+		s.timestampMap[pod.UID] = &podLocalTimestamps{CreationTimestamp: timeNow}
+	default:
+		s.timestampMap[pod.UID] = &podLocalTimestamps{CreationTimestamp: pod.CreationTimestamp.Time}
+	}
+	klog.V(4).InfoS("Process Pod Create Event", "Pod", klog.KObj(pod))
+}
+
+func (s *PodStore) onPodDelete(obj interface{}) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	timeNow := s.clock.Now()
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		var err error
+		pod, err = s.checkDeletedPod(obj)
+		if err != nil {
+			klog.ErrorS(err, "Got error while processing Delete Event")
+			return
+		}
+	}
+	timestamp, ok := s.timestampMap[pod.UID]
+	if !ok {
+		klog.ErrorS(nil, "Cannot find podLocalTimestamps in timestampMap", "UID", pod.UID)
+		return
+	}
+	timestamp.DeletionTimestamp = &timeNow
+	s.podsToDelete.AddAfter(pod, delayTime)
+	klog.V(4).InfoS("Process Pod Delete Event", "Pod", klog.KObj(pod))
+}
+
+func (s *PodStore) checkDeletedPod(obj interface{}) (*corev1.Pod, error) {
+	deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		return nil, fmt.Errorf("received unexpected object: %v", obj)
+	}
+	pod, ok := deletedState.Obj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("DeletedFinalStateUnknown object is not of type Pod: %v", deletedState.Obj)
+	}
+	return pod, nil
+}
+
+func (s *PodStore) GetPodByIPAndTime(ip string, startTime time.Time) (*corev1.Pod, bool) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	pods, _ := s.pods.ByIndex(podIPIndex, ip)
+	if len(pods) == 0 {
+		return nil, false
+	} else if len(pods) == 1 {
+		// In case the clocks may be skewed in different nodes, we directly return the Pod if there is only
+		// one Pod in the indexer. Otherwise, we check the timestamp for Pods in the indexer.
+		return pods[0].(*corev1.Pod), true
+	}
+	for _, pod := range pods {
+		pod := pod.(*corev1.Pod)
+		timestamp, ok := s.timestampMap[pod.UID]
+		if !ok {
+			continue
+		}
+		if timestamp.CreationTimestamp.Before(startTime) && (timestamp.DeletionTimestamp == nil || startTime.Before(*timestamp.DeletionTimestamp)) {
+			klog.V(4).InfoS("Get Pod info from indexer", "Pod", klog.KObj(pod))
```

Comment:
```suggestion
			klog.V(4).InfoS("Matched Pod IP and time to Pod from indexer", "ip", ip, "time", startTime, "Pod", klog.KObj(pod))
```

probably good to have a log message with more info here

Reply:
Updated